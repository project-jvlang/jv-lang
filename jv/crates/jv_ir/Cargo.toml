[package]
name = "jv_ir"
version.workspace = true
edition.workspace = true
license.workspace = true

[features]
default = []
debug-ir = []
perf-metrics = []

[[test]]
name = "perf_phase1"
path = "../../tests/performance/phase1.rs"

[dependencies]
anyhow = { workspace = true }
jv_ast.workspace = true
jv_parser.workspace = true
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
jv_support.workspace = true
ureq = { version = "2.9" }
url = "2.4"
sha2 = "0.10"
hex = "0.4"
tempfile = "3.10"
which = "4.4"
urlencoding = "2.1"
csv = "1.3"
