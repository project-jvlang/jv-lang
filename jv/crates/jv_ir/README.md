# jv_ir - Intermediate Representation for jv Language Desugaring

This crate handles the transformation of jv language constructs into desugared intermediate representation (IR) that can be easily compiled to Java 25 source code.

## Purpose

The IR serves as a clean intermediate layer between jv's syntax sugar and Java's more verbose constructs. It:

1. **Removes all jv sugar** while preserving semantics
2. **Provides explicit Java types** for all expressions and statements
3. **Enables straightforward Java code generation** without complex logic in the codegen phase
4. **Maintains source mapping** information for debugging

## Current Status: TDD Red Phase ✅

The crate is in the **Red phase** of Test-Driven Development:

- ✅ **Comprehensive IR node structure** designed for all jv language constructs
- ✅ **Complete type system** (`JavaType`) representing Java 25 types
- ✅ **Transformation context** with scope management and utility class tracking
- ✅ **34 comprehensive failing tests** covering all desugaring transformations
- ✅ **Error types** for all transformation scenarios
- ✅ **Public API** designed for integration with `jv_codegen_java`

All transformation functions correctly panic with "not yet implemented" messages, ready for the Green phase implementation.

## Desugaring Transformations Covered

### Variable Declarations
- `val x = 42` → `final int x = 42;`
- `var y: String?` → `String y = null;` (with null safety analysis)
- Type inference from initializers

### When Expressions
- `when (x) { 1 -> "one", 2 -> "two", else -> "other" }` → Java switch expressions
- Pattern matching with guards → case labels with conditions
- Range patterns → multiple case labels or guard conditions

### Extension Functions
- `fun String.reversed(): String` → Static method in utility class
- Generic extension functions → Parameterized static methods

### String Interpolation
- `"Hello, ${name}!"` → `String.format("Hello, %s!", name)` or concatenation

### Concurrency Constructs
- `spawn { ... }` → `Thread.ofVirtual().start(() -> { ... })`
- `async { ... }` → `CompletableFuture.supplyAsync(() -> { ... })`
- `future.await()` → `future.get()` or `future.join()`

### Resource Management
- `use(resource) { ... }` → try-with-resources statements
- `defer { ... }` → finally blocks

### Default Parameters
- Functions with default parameters → Multiple method overloads
- Named arguments → Proper parameter ordering

### Data Classes
- `data class User(val id: Int, var name: String)` → Java records or classes
- Immutable data classes → Java records
- Mutable data classes → Regular Java classes with getters/setters

### Null Safety
- `obj?.property` → Explicit null checks with conditional expressions
- `arr?[index]` → Null-safe array access
- `nullableValue ?: "default"` → Conditional expressions

### Top-level Functions
- Top-level functions → Static methods in utility classes
- Package-based utility class naming

## IR Node Structure

### Core Types
- **`JavaType`**: Represents all Java 25 types (primitives, references, arrays, functionals)
- **`IrExpression`**: Desugared expressions with explicit Java semantics
- **`IrStatement`**: Desugared statements including class/method declarations
- **`TransformContext`**: Maintains type information and transformation state

### Key Features
- **Explicit typing**: Every expression has a resolved Java type
- **Source preservation**: All nodes maintain span information
- **Java compatibility**: All constructs map directly to valid Java 25 code
- **Utility tracking**: Generated utility classes and method overloads are tracked

## Test Coverage

The test suite includes:
- **Basic transformation tests** for all jv constructs
- **Complex integration tests** for combined features
- **Error handling tests** for all error scenarios
- **Type system tests** for Java type compatibility
- **Context management tests** for scope tracking

## Integration Points

- **Input**: `jv_ast::Program` from the parser
- **Output**: `IrProgram` ready for Java code generation
- **Dependencies**: 
  - `jv_ast` for input AST nodes
  - `serde` for serialization support
  - `thiserror` for error handling

## Next Steps (Green Phase)

The Red phase is complete. The next phase will implement:

1. **Type inference engine** for inferring Java types from jv expressions
2. **Pattern desugaring** for when expressions to switch statements
3. **Null safety analysis** for safe nullable operations
4. **Extension function transformation** to static method calls
5. **Concurrency desugaring** to Java virtual threads and CompletableFuture
6. **Resource management** to try-with-resources and finally blocks
7. **Utility class generation** for top-level functions and extensions

## Architecture

The IR transformation follows a multi-pass approach:

1. **Analysis Pass**: Type inference and scope resolution
2. **Desugaring Pass**: Transform high-level constructs to basic operations
3. **Generation Pass**: Create utility classes and method overloads
4. **Validation Pass**: Ensure all IR nodes are valid for Java generation

The design enables incremental implementation and thorough testing of each transformation.