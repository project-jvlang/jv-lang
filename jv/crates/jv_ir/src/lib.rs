pub mod context;
pub mod error;
pub mod profiling;
pub mod sequence_pipeline;
pub mod transform;
pub mod types;

#[cfg(feature = "debug-ir")]
pub mod debug;

pub use jv_support::arena::{
    ArenaAccessor as TransformArenaAccessor, PoolMetrics as TransformPoolMetrics,
    PoolSessionMetrics as TransformPoolSessionMetrics, TransformPools, TransformPoolsGuard,
};

pub use profiling::{PerfMetrics, StageTiming, TransformProfiler};

/// Convenience re-exports for common IR types and transformation helpers.
pub mod prelude {
    pub use crate::context::TransformContext;
    pub use crate::error::TransformError;
    pub use crate::sequence_pipeline::*;
    pub use crate::transform::*;
    pub use crate::types::*;
    pub use crate::{
        PerfMetrics, StageTiming, TransformArenaAccessor, TransformPoolMetrics,
        TransformPoolSessionMetrics, TransformPools, TransformPoolsGuard, TransformProfiler,
    };
}

pub use prelude::*;

#[cfg(test)]
mod tests;
