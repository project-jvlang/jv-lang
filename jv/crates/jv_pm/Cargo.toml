[package]
name = "jv_pm"
version.workspace = true
edition.workspace = true
license.workspace = true
authors.workspace = true
repository.workspace = true

[dependencies]
thiserror.workspace = true
serde.workspace = true
serde_json.workspace = true
toml.workspace = true