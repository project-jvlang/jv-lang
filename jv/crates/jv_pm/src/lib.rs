// jv_pm - Package Manager functionality
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt;
use std::fs;
use std::path::Path;
use std::str::FromStr;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum PackageError {
    #[error("Package not found: {0}")]
    PackageNotFound(String),
    #[error("Version conflict: {0}")]
    VersionConflict(String),
    #[error("Invalid package manifest: {0}")]
    InvalidManifest(String),
    #[error("Network error: {0}")]
    NetworkError(String),
    #[error("IO error: {0}")]
    IoError(#[from] std::io::Error),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum JavaTarget {
    Java21,
    Java25,
}

impl JavaTarget {
    pub const fn as_str(self) -> &'static str {
        match self {
            JavaTarget::Java21 => "21",
            JavaTarget::Java25 => "25",
        }
    }

    pub fn variants() -> &'static [&'static str] {
        &["21", "25"]
    }

    pub const fn enables_modern_features(self) -> bool {
        matches!(self, JavaTarget::Java25)
    }

    pub const fn release_flag(self) -> &'static str {
        self.as_str()
    }
}

impl Default for JavaTarget {
    fn default() -> Self {
        JavaTarget::Java25
    }
}

impl fmt::Display for JavaTarget {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

#[derive(Debug, Clone)]
pub struct JavaTargetParseError {
    invalid: String,
}

impl JavaTargetParseError {
    fn new(value: impl Into<String>) -> Self {
        Self {
            invalid: value.into(),
        }
    }
}

impl fmt::Display for JavaTargetParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Unsupported java target '{}'. Supported targets: {}",
            self.invalid,
            JavaTarget::variants().join(", ")
        )
    }
}

impl std::error::Error for JavaTargetParseError {}

impl FromStr for JavaTarget {
    type Err = JavaTargetParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let normalised = s.trim().to_ascii_lowercase();
        match normalised.as_str() {
            "21" | "java21" | "jdk21" => Ok(JavaTarget::Java21),
            "25" | "java25" | "jdk25" | "latest" | "lts" => Ok(JavaTarget::Java25),
            other => Err(JavaTargetParseError::new(other)),
        }
    }
}

impl Serialize for JavaTarget {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(self.as_str())
    }
}

impl<'de> Deserialize<'de> for JavaTarget {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct Visitor;

        impl<'de> serde::de::Visitor<'de> for Visitor {
            type Value = JavaTarget;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("a Java target like '21' or '25'")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                JavaTarget::from_str(value).map_err(E::custom)
            }

            fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                self.visit_str(&value)
            }

            fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                JavaTarget::from_str(&value.to_string()).map_err(E::custom)
            }

            fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                if value < 0 {
                    return Err(E::custom(JavaTargetParseError::new(value.to_string())));
                }
                self.visit_u64(value as u64)
            }
        }

        deserializer.deserialize_any(Visitor)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PackageInfo {
    pub name: String,
    pub version: String,
    pub description: Option<String>,
    #[serde(default)]
    pub dependencies: HashMap<String, String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manifest {
    pub package: PackageInfo,
    #[serde(default)]
    pub project: ProjectSection,
    pub build: Option<BuildInfo>,
}

impl Manifest {
    pub fn load_from_path(path: impl AsRef<Path>) -> Result<Self, PackageError> {
        let path = path.as_ref();
        let content = fs::read_to_string(path)?;
        toml::from_str(&content).map_err(|error| PackageError::InvalidManifest(error.to_string()))
    }

    pub fn java_target(&self) -> JavaTarget {
        self.build
            .as_ref()
            .map(|info| info.java_version)
            .unwrap_or_default()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct BuildInfo {
    #[serde(default)]
    #[serde(rename = "java_version")]
    pub java_version: JavaTarget,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct ProjectSection {
    pub sources: SourceSection,
    pub output: OutputSection,
    pub entrypoint: Option<String>,
}

impl Default for ProjectSection {
    fn default() -> Self {
        Self {
            sources: SourceSection::default(),
            output: OutputSection::default(),
            entrypoint: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct SourceSection {
    pub include: Vec<String>,
    pub exclude: Vec<String>,
}

impl Default for SourceSection {
    fn default() -> Self {
        Self {
            include: vec!["src/**/*.jv".to_string()],
            exclude: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct OutputSection {
    #[serde(default = "default_output_directory")]
    pub directory: String,
    #[serde(default)]
    pub clean: bool,
}

impl Default for OutputSection {
    fn default() -> Self {
        Self {
            directory: default_output_directory(),
            clean: false,
        }
    }
}

fn default_output_directory() -> String {
    "target".to_string()
}

pub struct PackageManager {
    registry_url: String,
}

impl PackageManager {
    pub fn new(registry_url: String) -> Self {
        Self { registry_url }
    }

    pub fn install_package(&self, _name: &str, _version: &str) -> Result<(), PackageError> {
        // Placeholder implementation
        Ok(())
    }

    pub fn resolve_dependencies(
        &self,
        _manifest: &Manifest,
    ) -> Result<Vec<PackageInfo>, PackageError> {
        // Placeholder implementation
        Ok(vec![])
    }
}

#[cfg(test)]
mod tests;
