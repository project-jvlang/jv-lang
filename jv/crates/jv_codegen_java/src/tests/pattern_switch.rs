use super::*;
use insta::assert_snapshot;
use jv_ir::{
    IrCaseLabel, IrDeconstructionComponent, IrDeconstructionPattern, IrExpression, IrModifiers,
    IrParameter, IrProgram, IrRecordComponent, IrStatement, IrVisibility, JavaType,
};
use jv_pm::JavaTarget;

fn render_program(program: &IrProgram, target: JavaTarget) -> String {
    let mut generator = JavaCodeGenerator::with_config(JavaCodeGenConfig::for_target(target));
    let unit = generator
        .generate_compilation_unit(program)
        .expect("code generation");
    unit.to_source(&JavaCodeGenConfig::for_target(target))
}

fn snapshot_program(name: &str, program: IrProgram) {
    let java25 = render_program(&program, JavaTarget::Java25);
    let java21 = render_program(&program, JavaTarget::Java21);
    let combined = format!(
        "// ---- Java 25 ----\n{}\n\n// ---- Java 21 ----\n{}\n",
        java25.trim_end(),
        java21.trim_end()
    );
    assert_snapshot!(name, combined);
}

fn snapshot_java21_fallback(name: &str, summary: &str, program: IrProgram) {
    let java21 = render_program(&program, JavaTarget::Java21);
    let content = format!("{summary}\n{}\n", java21.trim_end());
    assert_snapshot!(name, content);
}

fn example1_program() -> IrProgram {
    let object_type = reference_type("Object");
    let string = string_type();
    let int = int_type();

    let discriminant = Box::new(ir_identifier("x", &object_type));

    let string_case = switch_case(
        vec![IrCaseLabel::TypePattern {
            type_name: "String".to_string(),
            variable: "value".to_string(),
            deconstruction: None,
        }],
        None,
        IrExpression::MethodCall {
            receiver: Some(Box::new(ir_identifier("value", &string))),
            method_name: "length".to_string(),
            java_name: None,
            resolved_target: None,
            args: vec![],
            argument_style: CallArgumentStyle::Comma,
            java_type: int.clone(),
            span: dummy_span(),
        },
    );

    let int_case = switch_case(
        vec![IrCaseLabel::TypePattern {
            type_name: "Integer".to_string(),
            variable: "value".to_string(),
            deconstruction: None,
        }],
        None,
        IrExpression::Binary {
            left: Box::new(ir_identifier("value", &int)),
            op: BinaryOp::Multiply,
            right: Box::new(IrExpression::Literal(
                Literal::Number("2".to_string()),
                None,
                dummy_span(),
            )),
            java_type: int.clone(),
            span: dummy_span(),
        },
    );

    let default_case = switch_case(
        vec![IrCaseLabel::Default],
        None,
        IrExpression::Literal(Literal::Number("0".to_string()), None, dummy_span()),
    );

    let switch_expr = IrExpression::Switch {
        discriminant,
        cases: vec![string_case, int_case, default_case],
        java_type: int.clone(),
        implicit_end: None,
        strategy_description: Some(
            "strategy=Switch arms=3 guards=0 default=true exhaustive=false".to_string(),
        ),
        span: dummy_span(),
    };

    let method_body = IrExpression::Block {
        statements: vec![IrStatement::Return {
            value: Some(switch_expr),
            span: dummy_span(),
        }],
        java_type: int,
        span: dummy_span(),
    };

    let evaluate_method = IrStatement::MethodDeclaration {
        name: "evaluate".to_string(),
        java_name: None,
        type_parameters: vec![],
        parameters: vec![IrParameter {
            name: "x".to_string(),
            java_type: object_type,
            modifiers: IrModifiers::default(),
            span: dummy_span(),
        }],
        primitive_return: None,
        return_type: int_type(),
        body: Some(method_body),
        modifiers: IrModifiers {
            visibility: IrVisibility::Public,
            is_static: true,
            ..IrModifiers::default()
        },
        throws: vec![],
        span: dummy_span(),
    };

    let class = IrStatement::ClassDeclaration {
        name: "Example1".to_string(),
        type_parameters: vec![],
        superclass: None,
        interfaces: vec![],
        fields: vec![],
        methods: vec![evaluate_method],
        nested_classes: vec![],
        modifiers: IrModifiers {
            visibility: IrVisibility::Public,
            ..IrModifiers::default()
        },
        span: dummy_span(),
    };

    IrProgram {
        package: Some("patterns".to_string()),
        imports: vec![],
        type_declarations: vec![class],
        generic_metadata: Default::default(),
        conversion_metadata: Vec::new(),
        span: dummy_span(),
    }
}

fn example2_program() -> IrProgram {
    let int = int_type();

    fn range_case(lower: &str, upper: &str, inclusive: bool, label: &str) -> IrSwitchCase {
        let guard = if inclusive {
            BinaryOp::LessEqual
        } else {
            BinaryOp::Less
        };

        let lower_check = IrExpression::Binary {
            left: Box::new(ir_identifier("score", &int_type())),
            op: BinaryOp::GreaterEqual,
            right: Box::new(IrExpression::Literal(
                Literal::Number(lower.to_string()),
                None,
                dummy_span(),
            )),
            java_type: JavaType::boolean(),
            span: dummy_span(),
        };

        let upper_check = IrExpression::Binary {
            left: Box::new(ir_identifier("score", &int_type())),
            op: guard,
            right: Box::new(IrExpression::Literal(
                Literal::Number(upper.to_string()),
                None,
                dummy_span(),
            )),
            java_type: JavaType::boolean(),
            span: dummy_span(),
        };

        let condition = IrExpression::Binary {
            left: Box::new(lower_check),
            op: BinaryOp::And,
            right: Box::new(upper_check),
            java_type: JavaType::boolean(),
            span: dummy_span(),
        };

        switch_case(
            vec![IrCaseLabel::Range {
                type_name: "int".to_string(),
                variable: "it0".to_string(),
                lower: Box::new(IrExpression::Literal(
                    Literal::Number(lower.to_string()),
                    None,
                    dummy_span(),
                )),
                upper: Box::new(IrExpression::Literal(
                    Literal::Number(upper.to_string()),
                    None,
                    dummy_span(),
                )),
                inclusive_end: inclusive,
            }],
            Some(condition),
            IrExpression::Literal(Literal::String(label.to_string()), None, dummy_span()),
        )
    }

    let discriminant = Box::new(ir_identifier("score", &int));

    let cases = vec![
        range_case("0", "60", false, "Fail"),
        range_case("60", "80", false, "Pass"),
        range_case("80", "100", true, "Excellent"),
        switch_case(
            vec![IrCaseLabel::Default],
            None,
            IrExpression::Literal(Literal::String("Invalid".to_string()), None, dummy_span()),
        ),
    ];

    let switch_expr = IrExpression::Switch {
        discriminant,
        cases,
        java_type: string_type(),
        implicit_end: None,
        strategy_description: Some(
            "strategy=Switch arms=4 guards=3 default=true exhaustive=false".to_string(),
        ),
        span: dummy_span(),
    };

    let method_body = IrExpression::Block {
        statements: vec![IrStatement::Return {
            value: Some(switch_expr),
            span: dummy_span(),
        }],
        java_type: string_type(),
        span: dummy_span(),
    };

    let method = IrStatement::MethodDeclaration {
        name: "categorize".to_string(),
        java_name: None,
        type_parameters: vec![],
        parameters: vec![IrParameter {
            name: "score".to_string(),
            java_type: int_type(),
            modifiers: IrModifiers::default(),
            span: dummy_span(),
        }],
        primitive_return: None,
        return_type: string_type(),
        body: Some(method_body),
        modifiers: IrModifiers {
            visibility: IrVisibility::Public,
            is_static: true,
            ..IrModifiers::default()
        },
        throws: vec![],
        span: dummy_span(),
    };

    let class = IrStatement::ClassDeclaration {
        name: "Example2".to_string(),
        type_parameters: vec![],
        superclass: None,
        interfaces: vec![],
        fields: vec![],
        methods: vec![method],
        nested_classes: vec![],
        modifiers: IrModifiers {
            visibility: IrVisibility::Public,
            ..IrModifiers::default()
        },
        span: dummy_span(),
    };

    IrProgram {
        package: Some("patterns".to_string()),
        imports: vec![],
        type_declarations: vec![class],
        generic_metadata: Default::default(),
        conversion_metadata: Vec::new(),
        span: dummy_span(),
    }
}

fn example5_program() -> IrProgram {
    let int = int_type();
    let inner_type = reference_type("Inner");
    let outer_type = reference_type("Outer");

    let outer_record = IrStatement::RecordDeclaration {
        name: "Outer".to_string(),
        type_parameters: vec![],
        components: vec![
            record_component("inner", inner_type.clone()),
            record_component("count", int.clone()),
        ],
        interfaces: vec![],
        methods: vec![],
        modifiers: public_modifiers(),
        span: dummy_span(),
    };

    let inner_record = IrStatement::RecordDeclaration {
        name: "Inner".to_string(),
        type_parameters: vec![],
        components: vec![
            record_component("x", int.clone()),
            record_component("y", int.clone()),
        ],
        interfaces: vec![],
        methods: vec![],
        modifiers: public_modifiers(),
        span: dummy_span(),
    };

    let nested_pattern = IrDeconstructionPattern {
        components: vec![
            IrDeconstructionComponent::Type {
                type_name: "Inner".to_string(),
                pattern: Some(Box::new(IrDeconstructionPattern {
                    components: vec![
                        IrDeconstructionComponent::Binding {
                            name: "x".to_string(),
                        },
                        IrDeconstructionComponent::Binding {
                            name: "y".to_string(),
                        },
                    ],
                })),
            },
            IrDeconstructionComponent::Binding {
                name: "count".to_string(),
            },
        ],
    };

    let sum_xy = IrExpression::Binary {
        left: Box::new(ir_identifier("x", &int)),
        op: BinaryOp::Add,
        right: Box::new(ir_identifier("y", &int)),
        java_type: int.clone(),
        span: dummy_span(),
    };

    let total_body = IrExpression::Binary {
        left: Box::new(sum_xy),
        op: BinaryOp::Add,
        right: Box::new(ir_identifier("count", &int)),
        java_type: int.clone(),
        span: dummy_span(),
    };

    let match_case = switch_case(
        vec![IrCaseLabel::TypePattern {
            type_name: "Outer".to_string(),
            variable: "outer".to_string(),
            deconstruction: Some(nested_pattern),
        }],
        None,
        total_body,
    );

    let default_case = switch_case(vec![IrCaseLabel::Default], None, number_literal("0"));

    let switch_expr = IrExpression::Switch {
        discriminant: Box::new(ir_identifier("value", &outer_type)),
        cases: vec![match_case, default_case],
        java_type: int.clone(),
        implicit_end: None,
        strategy_description: Some(
            "strategy=Switch arms=2 guards=0 default=true exhaustive=true".to_string(),
        ),
        span: dummy_span(),
    };

    let return_stmt = IrStatement::Return {
        value: Some(switch_expr),
        span: dummy_span(),
    };

    let total_method = IrStatement::MethodDeclaration {
        name: "total".to_string(),
        java_name: None,
        type_parameters: vec![],
        parameters: vec![IrParameter {
            name: "value".to_string(),
            java_type: outer_type,
            modifiers: IrModifiers::default(),
            span: dummy_span(),
        }],
        primitive_return: None,
        return_type: int.clone(),
        body: Some(IrExpression::Block {
            statements: vec![return_stmt],
            java_type: int.clone(),
            span: dummy_span(),
        }),
        modifiers: method_modifiers(),
        throws: vec![],
        span: dummy_span(),
    };

    let example_class = IrStatement::ClassDeclaration {
        name: "Example5".to_string(),
        type_parameters: vec![],
        superclass: None,
        interfaces: vec![],
        fields: vec![],
        methods: vec![total_method],
        nested_classes: vec![],
        modifiers: public_modifiers(),
        span: dummy_span(),
    };

    IrProgram {
        package: Some("patterns".to_string()),
        imports: vec![],
        type_declarations: vec![inner_record, outer_record, example_class],
        generic_metadata: Default::default(),
        conversion_metadata: Vec::new(),
        span: dummy_span(),
    }
}

fn method_modifiers() -> IrModifiers {
    IrModifiers {
        visibility: IrVisibility::Public,
        is_static: true,
        ..IrModifiers::default()
    }
}

fn public_modifiers() -> IrModifiers {
    IrModifiers {
        visibility: IrVisibility::Public,
        ..IrModifiers::default()
    }
}

fn record_component(name: &str, java_type: JavaType) -> IrRecordComponent {
    IrRecordComponent {
        name: name.to_string(),
        java_type,
        span: dummy_span(),
    }
}

fn number_literal(value: &str) -> IrExpression {
    IrExpression::Literal(Literal::Number(value.to_string()), None, dummy_span())
}

#[test]
fn pattern_example1_type_patterns() {
    snapshot_program("pattern_example1_type_patterns", example1_program());
}

#[test]
fn pattern_example2_range_patterns() {
    snapshot_program("pattern_example2_range_patterns", example2_program());
}

#[test]
fn pattern_example5_destructuring_patterns() {
    snapshot_program(
        "pattern_example5_destructuring_patterns",
        example5_program(),
    );
}

#[test]
fn pattern_example1_java21_fallback_instanceof() {
    snapshot_java21_fallback(
        "pattern_example1_java21_fallback_instanceof",
        "// Java 21 fallback expands type patterns into instanceof chains guarded by __matched",
        example1_program(),
    );
}

#[test]
fn pattern_example2_java21_fallback_ranges() {
    snapshot_java21_fallback(
        "pattern_example2_java21_fallback_ranges",
        "// Java 21 fallback lowers range patterns into sequential guard checks with __matched",
        example2_program(),
    );
}

#[test]
fn pattern_example5_java21_fallback_depth2() {
    snapshot_java21_fallback(
        "pattern_example5_java21_fallback_depth2",
        "// Java 21 fallback handles nested destructuring via instanceof loop and do-while guard",
        example5_program(),
    );
}
