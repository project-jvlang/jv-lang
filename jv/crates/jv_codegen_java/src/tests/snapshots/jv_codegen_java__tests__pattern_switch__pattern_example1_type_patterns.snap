---
source: crates/jv_codegen_java/src/tests/pattern_switch.rs
expression: combined
---
// ---- Java 25 ----
package patterns;

public class Example1 {
    public static int evaluate(Object x) {
        return // strategy=Switch arms=3 guards=0 default=true exhaustive=false
        switch (x) {
            case String value -> value.length();
            case Integer value -> value * 2;
            default -> 0;
        }
        ;
    }
    
}

// ---- Java 21 ----
package patterns;

public class Example1 {
    public static int evaluate(Object x) {
        return // strategy=Switch arms=3 guards=0 default=true exhaustive=false
        new Object() {
            int matchExpr() {
                final var __subject = x;
                final int __matchResult;
                boolean __matched = false;
                if (!__matched) {
                    if (__subject instanceof String value) {
                        __matchResult = value.length();
                        __matched = true;
                    }
                }
                if (!__matched) {
                    if (__subject instanceof Integer value) {
                        __matchResult = value * 2;
                        __matched = true;
                    }
                }
                if (!__matched) {
                    __matchResult = 0;
                    __matched = true;
                }
                return __matchResult;
            }
        }.matchExpr()
        ;
    }
    
}
